### PR TITLE
Implement `OffchainSignatures` module - universal storage for public keys and params of different signature schemes used offchain

### PR DESCRIPTION
**Pull request by olegnn**
- Implement `OffchainSignatures` module (will replace current `BbsPlusModule`) to store offchain signature keys and parameters for different signature schemes. Currently can be either `BBS`, `BBS+` or `Pointcheval-Sanders`.
- Remove previous migrations.

### CODE REVIEW

File: pallets/core_mods/src/modules/offchain_signatures/mod.rs [hunk 1]
```diff
@@ -46,8 +47,10 @@ pub trait Config: system::Config + did::Config {
     type ParamsMaxSize: Get<u32>;
     /// Weight consumed per byte of the params. This will determine the cost of the transaction.
     type ParamsPerByteWeight: Get<Weight>;
-    /// Maximum byte size of the public key. This depends only on the chosen elliptic curve.
-    type PublicKeyMaxSize: Get<u32>;
+    /// Maximum byte size of the `BBS`/`BBS+` (fixed) public key. This depends only on the chosen elliptic curve.
+    type BBSPublicKeyMaxSize: Get<u32>;
+    /// Maximum byte size of the `PS` public key. This depends only on the chosen elliptic curve.
```
Comment (lovesh) [inline pallets/core_mods/src/modules/offchain_signatures/mod.rs:52]: This also depends on the number of messages supported.

File: pallets/core_mods/src/modules/offchain_signatures/actions.rs [hunk 2]
```diff
@@ -1,51 +1,51 @@
 use super::*;
 
-#[derive(Encode, Decode, scale_info_derive::TypeInfo, Clone, PartialEq, Debug)]
+#[derive(Encode, Decode, scale_info_derive::TypeInfo, Clone, PartialEq, Eq, Debug)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
 #[scale_info(skip_type_params(T))]
 #[scale_info(omit_prefix)]
```
Comment (lovesh) [inline pallets/core_mods/src/modules/offchain_signatures/actions.rs:6]: What is `omit_prefix` for?

File: pallets/core_mods/src/modules/offchain_signatures/schemas.rs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            pub(crate) curve_type: CurveType,
+            pub(crate) bytes: Bytes,
+            /// The params used to generate the public key
+            pub(crate) params_ref: Option<SignatureParamsStorageKey>,
+            /// Optional participant id used in threshold issuance.
+            pub(crate) participant_id: Option<u16>,
+        }
+
+        impl $key {
+            /// Instantiates new public key for the BBS+ signature scheme.
```
Comment (lovesh) [inline pallets/core_mods/src/modules/offchain_signatures/schemas.rs:29]: `key for the BBS+ signature` -> `key for the signature`

File: pallets/core_mods/src/modules/offchain_signatures/schemas.rs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+                    params_ref: params_ref.into(),
+                    curve_type,
+                    participant_id: None,
+                }
+            }
+
+            /// Instantiates new public key with participant id for the BBS+ signature scheme.
+            /// This function doesn't validate supplied bytes.
+            /// Participant id implies the usage of this key in threshold issuance.
+            pub fn new_participant(
```
Comment (lovesh) [inline pallets/core_mods/src/modules/offchain_signatures/schemas.rs:47]: `fn new_participant` -> `fn new_with_participant_id`

File: pallets/core_mods/src/modules/offchain_signatures/schemas.rs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        #[scale_info(omit_prefix)]
+        pub struct $key {
+            /// The public key should be for the same curve as the parameters but a public key might not have
+            /// parameters on chain
+            pub(crate) curve_type: CurveType,
+            pub(crate) bytes: Bytes,
+            /// The params used to generate the public key
+            pub(crate) params_ref: Option<SignatureParamsStorageKey>,
+            /// Optional participant id used in threshold issuance.
+            pub(crate) participant_id: Option<u16>,
```
Comment (lovesh) [inline pallets/core_mods/src/modules/offchain_signatures/schemas.rs:25]: `u16` -> `ParticipantId` like crypto lib.

File: pallets/core_mods/src/modules/offchain_signatures/schemas.rs [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+            ) -> Self {
+                Self {
+                    bytes: bytes.into(),
+                    params_ref: params_ref.into(),
+                    curve_type,
+                    participant_id: None,
+                }
+            }
+
+            /// Instantiates new public key with participant id for the BBS+ signature scheme.
```
Comment (lovesh) [inline pallets/core_mods/src/modules/offchain_signatures/schemas.rs:44]: Similar comment as above about scheme name

**Review by lovesh — CHANGES_REQUESTED** [review]

File: pallets/core_mods/src/modules/offchain_signatures/actions.rs (hunk 2 above)
Comment (olegnn) [reply pallets/core_mods/src/modules/offchain_signatures/actions.rs:6]: To omit the default prefix created by the `scale_info` (file path).

File: pallets/core_mods/src/modules/offchain_signatures/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (olegnn) [reply pallets/core_mods/src/modules/offchain_signatures/mod.rs:52]: Comment for the message count-dependent params size limit was the following: `Maximum byte size of the parameters. This depends only on the chosen elliptic curve.`.
I fixed both.

File: pallets/core_mods/src/modules/offchain_signatures/schemas.rs [hunk 7]
```diff
@@ -0,0 +1,176 @@
+use crate::{offchain_signatures::SignatureParams, types::CurveType, util::Bytes};
```
Comment (lovesh) [inline pallets/core_mods/src/modules/offchain_signatures/schemas.rs:1]: Change the file name and other occurrences of the word "schema" (in this context) to "scheme".

**Pull request merged**